### ISSUE TITLE
Windows failure with PR 46541

### ISSUE THREAD

**Issue by whytewolf** [saltstack/salt#46906]
### Description of Issue/Question

Using the new version of metadata.py from https://github.com/saltstack/salt/pull/46541. Works in linux. However, depending on if using py2 or py3 casts an error in windows.

In py2 windows error is as follows:

```
2018-04-05 02:52:35,470 [salt.loader      :763 ][CRITICAL][2496] Failed to load grains defined in grain file metadata.metadata in function <function metadata at 0x0000000001BA9198>, error:
Traceback (most recent call last):
  File "c:\salt\bin\lib\site-packages\salt\loader.py", line 756, in grains
    ret = funcs[key]()
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 79, in metadata
    return _search()
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 63, in _search
    ret[line] = _search(prefix=os.path.join(prefix, line + '/'))
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 58, in _search
    ret[line[:-1]] = _search(prefix=os.path.join(prefix, line))
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 66, in _search
    ret[value] = _search(prefix=os.path.join(prefix, key))
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 72, in _search
    ret[line] = json.loads(retdata)
  File "c:\salt\bin\lib\json\__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "c:\salt\bin\lib\json\decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
TypeError: expected string or buffer
```

In py3 windows error is as follows:
```
[CRITICAL] Failed to load grains defined in grain file metadata.metadata in function <function metadata at 0x04EB96A8>,
error:
Traceback (most recent call last):
  File "C:\salt\bin\lib\site-packages\salt\loader.py", line 756, in grains
    ret = funcs[key]()
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 79, in metadata
    return _search()
  File "c:\salt\var\cache\salt\minion\extmods\grains\metadata.py", line 56, in _search
    for line in linedata['body'].split('\n'):
TypeError: a bytes-like object is required, not 'str'
```

### Setup
Windows running salt with metadata source from https://github.com/saltstack/salt/pull/46541

### Versions Report
```
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: 1.10.0
       cherrypy: 10.2.1
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: 1.6.1
       M2Crypto: Not Installed
           Mako: 1.0.7
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:19:30) [MSC v.1500 32 bit (Intel)]
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 16.0.3
           RAET: Not Installed
          smmap: 2.0.3
        timelib: 0.2.4
        Tornado: 4.5.2
            ZMQ: 4.1.6

System Versions:
           dist:
         locale: cp1252
        machine: AMD64
        release: 2016Server
         system: Windows
        version: 2016Server 10.0.14393  Multiprocessor Free
```

```
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: 1.10.0
       cherrypy: 10.2.1
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.7
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.5.3 (v3.5.3:1880cb95a742, Jan 16 2017, 15:51:26) [MSC v.1900 32 bit (Intel)]
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 16.0.3
           RAET: Not Installed
          smmap: 2.0.3
        timelib: 0.2.4
        Tornado: 4.5.2
            ZMQ: 4.1.6

System Versions:
           dist:
         locale: cp1252
        machine: AMD64
        release: 2016Server
         system: Windows
        version: 2016Server 10.0.14393 SP0 Multiprocessor Free
```

**Comment by whytewolf** [comment]
ZD-2405

**Comment by gtmanfred** [comment]
Can you test this PR?

https://github.com/saltstack/salt/pull/46541

It will be in 2017.7.6 and 2018.3.1

**Comment by gtmanfred** [comment]
Ok, i will test windows.

Thanks,
Daniel

**Comment by gtmanfred** [comment]
This should be fixed by https://github.com/saltstack/salt/pull/47168

**Comment by rallytime** [comment]
Closed via #47168
